### PR TITLE
Cache bytecode recording proxies

### PR DESCRIPTION
**Pull request by stuartwdouglas**

### CODE REVIEW

**Comment by gsmet** [comment]
Failing with:
```
[ERROR] Tests run: 18, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.952 s <<< FAILURE! - in io.quarkus.deployment.recording.BytecodeRecorderTestCase
[ERROR] testLargeNumberOfInvocations  Time elapsed: 0.319 s  <<< ERROR!
java.util.ConcurrentModificationException
	at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:909)
	at java.util.ArrayList$Itr.next(ArrayList.java:859)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.writeBytecode(BytecodeRecorderImpl.java:416)
	at io.quarkus.deployment.recording.BytecodeRecorderTestCase.runTest(BytecodeRecorderTestCase.java:300)
	at io.quarkus.deployment.recording.BytecodeRecorderTestCase.testLargeNumberOfInvocations(BytecodeRecorderTestCase.java:223)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:686)
```

**Comment by stuartwdouglas** [comment]
Failure look unrelated.

File: core/deployment/src/main/java/io/quarkus/deployment/recording/BytecodeRecorderImpl.java [hunk 1]
```diff
@@ -224,60 +227,66 @@ private static boolean isProxiable(Class<?> returnType) {
...
+                return instance.proxy;
+            }
+
+        };
+
+        try {
+            if (recordingProxyFactories.containsKey(theClass)) {
+                return (T) recordingProxyFactories.get(theClass).newInstance(invocationHandler);
+            }
+            String proxyNameSuffix = "$$RecordingProxyProxy" + COUNT.incrementAndGet();
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/recording/BytecodeRecorderImpl.java:279]: It's minor but wouldn't we want a stable name here too?

Comment (stuartwdouglas) [reply core/deployment/src/main/java/io/quarkus/deployment/recording/BytecodeRecorderImpl.java:279]: It does not really matter here, these don't go into the final application, and as all they do is record bytecode they don't show up in any profiles (as this is basically just adding to a list).

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
I understand this is a performance improvement. I'm more comfortable keeping that one for 1.5 and make a big splash with all the improvements you made.

**Comment by stuartwdouglas** [comment]
The finalize problem could actually happen to current applications I think,
so there is also a bug fix component

**Comment by gsmet** [comment]
Ok. I added the tag back.

**Pull request merged**